src/frame/mod.rs

```diff
@@ -54,29 +75,53 @@ impl fmt::Display for Error {
 
 impl std::error::Error for Error {}
 
-/// Compress all bytes of `input`.
-pub fn compress(input: &[u8]) -> Vec<u8> {
+/// Compress all bytes of `input` into a `Vec`.
+pub fn compress(input: &[u8]) -> Result<Vec<u8>, Error> {
     compress_with(FrameInfo::default(), input)
 }
 
-/// Compress all bytes of `input`.
-pub fn compress_with(frame_info: FrameInfo, input: &[u8]) -> Vec<u8> {
+/// Compress all bytes of `input` into a `Vec` with the specified Frame configuration.
+pub fn compress_with(frame_info: FrameInfo, input: &[u8]) -> Result<Vec<u8>, Error> {
     let buffer = Vec::with_capacity(
         header::MAX_FRAME_INFO_SIZE
             + header::BLOCK_INFO_SIZE
             + crate::block::compress::get_maximum_output_size(input.len()),
     );
-    let mut enc = FrameEncoder::with_frame_info(frame_info, buffer);
-    enc.write_all(input).unwrap();
-    enc.finish().unwrap()
+    let mut enc = FrameEncoder::with_frame_info(frame_info, buffer)?;
+    enc.write_all(input)?;
+    Ok(enc.finish()?)
 }
 
 /// Decompress all bytes of `input` into a new vec.
 pub fn decompress(input: &[u8]) -> Result<Vec<u8>, Error> {
     let mut de = FrameDecoder::new(input);
-    // Preallocate the Vec with 1.5x the size of input, it may resize but it amortizes enough.
+    // Preallocate the output with 2x the input size, it may resize but it amortizes enough.
     // The upside is that we don't have to worry about DOS attacks, etc..
-    let mut out = Vec::with_capacity(input.len() * 3 / 2);
+    let mut out = Vec::with_capacity(input.len() * 2);
```

Comment:
I think the compression can't be that bad? The maximum overhead should be much lower

Reply:
I'm not sure I understand what you mean. It's assuming a 0.5 compression ratio. Which is reasonable, right?

Reply:
ah I was thinking this preallocates for compression.
In the decompression case I don't know how this can affect DOS. I think for preallocation we should try to read the content size

Reply:
That's how you DOS. Build a valid frame with a massive content size.

Reply:
I considered doing that with a maximum limit to the allocation, but it required the specialisation feature. So when the  Frame Decoder read the FrameInfo it would try preallocating the output (with the specialisation trait).

Then I considered injecting an already read FrameInfo which seemed more doable but still significant tinkering with FrameDecoder (as it would not be reading the FrameInfo itself).

I ended up giving up and opting for the estimation based on the input size.

Reply:
What I was trying to say is that reading content length is very prone to DOS a naive decoder will preallocate `content_size`  bytes after only reading 15 bytes.

> If the frame is massive we will keep reading until the system runs out of Ram/disk space etc..
I see your point. It's still possible if the attacker can provide enough input, this one is not worth worrying about though.

Reply:
> 
> I see your point. It's still possible if the attacker can provide enough input, this one is not worth worrying about though.

With lz4 you can have a maximum ratio of around 250. Still a valid attack vector, but yes it's more difficult.